flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/filemapping/FileMappingManager.java

```diff
@@ -0,0 +1,252 @@
+            sourceEntry.retain();
+            mappingTable.put(src, sourceEntry);
+            sourceEntry.retain();
+            mappingTable.put(dst, sourceEntry);
+        }
+        LOG.trace("link: {} -> {}", dst, src);
+        return 0;
+    }
+
+    /** Get the real path of a file, the real path maybe a local file or a remote file/dir. */
+    public RealPath realPath(Path file) {
+        String fileName = file.toString();
+        MappingEntry entry = mappingTable.getOrDefault(fileName, null);
+        if (!fileName.endsWith(SST_SUFFIX) && fileName.startsWith(remoteBase)) {
+            if (entry != null) {
+                return new RealPath(new Path(entry.sourcePath), true);
+            }
+            Path localFile = new Path(localBase, file.getName());
+            return new RealPath(localFile, true);
+        }
+        if (entry != null) {
+            return new RealPath(new Path(entry.sourcePath), false);
+        }
+        return new RealPath(file, false);
+    }
+
+    public List<String> listByPrefix(String path) {
+        List<String> linkedPaths = new ArrayList<>();
+        for (Map.Entry<String, MappingEntry> entry : mappingTable.entrySet()) {
+            if (entry.getKey().startsWith(path) && !entry.getValue().isDirectory) {
+                linkedPaths.add(entry.getKey());
+            }
+        }
+        return linkedPaths;
+    }
+
+    /**
+     * 1. If src can match any key, we only `mark rename`, no physical file would be renamed. 2. If
+     * src is a directory, all files under src will be renamed, including linked files and local
+     * files, the directory also would be renamed in file system physically.
+     *
+     * @param src the source path
+     * @param dst the destination path
+     * @return always return true except for IOException
+     */
+    public boolean renameFile(String src, String dst) throws IOException {
+        MappingEntry srcEntry = mappingTable.get(src);
+        if (srcEntry != null) { // rename file
+            if (mappingTable.containsKey(dst)) {
+                MappingEntry dstEntry = mappingTable.remove(dst);
+                dstEntry.release();
+            }
+            mappingTable.remove(src);
+            mappingTable.put(dst, srcEntry);
+        } else { // rename directory
+            // 1. rename file in old directory
+            List<String> toRename = new ArrayList<>();
+            for (String key : mappingTable.keySet()) {
+                if (isParentDir(key, src) || key.equals(src)) {
+                    toRename.add(key);
+                }
+            }
+            if (!toRename.isEmpty()) {
+                for (String key : toRename) {
+                    MappingEntry sourceEntry = mappingTable.remove(key);
+                    String renamedDst = key.replace(src, dst);
+                    LOG.trace("rename: {} -> {}", key, renamedDst);
+                    sourceEntry.sourcePath = sourceEntry.sourcePath.replace(src, dst);
```

Comment:
Why the `sourceEntry.sourcePath` changed?

Reply:
After discussing offline, we use `link to dst dir + delete src dir` implement `rename`, the changes of `sourceEntry.sourcePath` have been removed.